stroyprombeton/migrations/0019_catalog_data.py

```diff
@@ -0,0 +1,96 @@
+# -*- coding: utf-8 -*-
+# Generated by Django 1.11.20 on 2019-03-04 15:21
+from __future__ import unicode_literals
+
+from collections import defaultdict
+from functools import lru_cache
+
+from django.db import migrations
+
+APP_LABEL = 'stroyprombeton'
+
+FIELD_TO_TAG_GROUP = {
+    'specification': 'Серия',
+    'length': 'Длина',
+    'width': 'Ширина',
+    'height': 'Высота',
+    'weight': 'Масса',
+    'volume': 'Объём',
+    'diameter_out': 'Внешний диаметр',
+    'diameter_in': 'Внутренний диаметр',
+}
+
+TAG_FIELDS = FIELD_TO_TAG_GROUP.keys()
+
+OPTION_FIELDS = [
+    'date_price_updated', 'code', 'mark', 'is_new_price', 'price', 'in_stock',
+]
+
+
+@lru_cache(maxsize=1)
+def group_products(model) -> dict:
+    grouped = defaultdict(list)
+    for product in model.objects.order_by('id'):
+        grouped[(product.name, product.category.id)].append(product)
+    return grouped
+
+
+def create_options(apps, schema_editor):
+    product_model = apps.get_model(APP_LABEL, 'Product')
+    option_model = apps.get_model(APP_LABEL, 'Option')
+
+    for (name, _), products in group_products(product_model).items():
+        # the product with the lowest in the groups will be only left in DB
+        main_product = products[0]
+        option_model.objects.bulk_create(
+            [option_model(
+                product=main_product,
+                **{f: getattr(product, f) for f in OPTION_FIELDS},
+                **{f: getattr(product, f) for f in TAG_FIELDS}
+            ) for product in products]
+        )
+
+
+def clean_products(apps, schema_editor):
+    """Safely remove product duplicates."""
+    product_model = apps.get_model(APP_LABEL, 'Product')
+
+    for (name, _), products in group_products(product_model).items():
+        product_model.objects.filter(
+            id__in=[p.id for p in products[1:]]
+        ).delete()
+
+
+def create_tags(apps, schema_editor):
+    """Move data from option columns to tag records."""
+    group_model = apps.get_model(APP_LABEL, 'TagGroup')
+    tag_model = apps.get_model(APP_LABEL, 'Tag')
+    option_model = apps.get_model(APP_LABEL, 'Option')
+
+    field_to_group = {
+        field: group_model(name=group, position=pos)
+        for pos, (field, group) in enumerate(FIELD_TO_TAG_GROUP.items())
+    }
+    for group in field_to_group.values():
+        group.save()
+
+    for option in option_model.objects.all():
+        for field, group in field_to_group.items():
+            value = getattr(option, field)
+            if value:
+                tag, _ = tag_model.objects.get_or_create(name=value, group=group)
+                tag.options.add(option)
```

Comment:
```suggestion
                option.tags.add(tag)
```
We save `option`, so we should modify option